### PR TITLE
#1736 race condition in the federated sail

### PR DESCRIPTION
**Pull request by hmottestad**
GitHub issue resolved: #1736   

Briefly describe the changes proposed in this PR:

* reproduce bug consistently
* fix race condition with synchronization

### CODE REVIEW

**Comment by hmottestad** [comment]
@jeenbroekstra this is reliably reproducing the test error on my machine in `SPARQLBuilderTest`.

**Comment by jeenbroekstra** [comment]
@aschwarte10 do you have an opinion on this? I am happy to disable this test for the time being, pending deprecation of the federation sail, but perhaps you see an obvious cause?

File: core/sail/federation/src/main/java/org/eclipse/rdf4j/sail/federation/evaluation/ParallelJoinCursor.java [hunk 1]
```diff
@@ -75,11 +76,17 @@ public ParallelJoinCursor(EvaluationStrategy strategy,
...
+				synchronized (this) {
+					if (!closed && !isClosed() && leftIter.hasNext()) {
+						CloseableIteration<BindingSet, QueryEvaluationException> evaluate = strategy.evaluate(rightArg,
+								leftIter.next());
+						toCloseList.add(evaluate);
+						rightQueue.put(evaluate);
+					} else {
+						break;
+					}
+				}
```
Comment (hmottestad) [inline core/sail/federation/src/main/java/org/eclipse/rdf4j/sail/federation/evaluation/ParallelJoinCursor.java]: The problem was that between checking for closed and hasNext the iteration could have been closed. So I introduced some light locking in order to eliminate this race condition.

**Comment by hmottestad** [comment]
@jeenbroekstra got this one figured out. It was a race condition in how the code detects when the iteration has been closed. Got it fixed. Ready to merge.

**Pull request merged**